File: server/engines/scm-ai.js
/**
 * TrustChecker Supply Chain AI Engine
 * Statistical analysis for delay prediction, route optimization,
 * inventory forecast, bottleneck detection, and partner risk scoring
 */

class ScmAI {
    /**
     * Predict delivery delay based on historical shipment data
     * Uses exponential weighted moving average
     */
    predictDelay(shipments) {
        if (!shipments || shipments.length < 2) return { predicted_delay_hours: 0, confidence: 0.5, risk: 'low' };

        const delays = shipments
            .filter(s => s.actual_delivery && s.estimated_delivery)
            .map(s => {
                const est = new Date(s.estimated_delivery).getTime();
                const act = new Date(s.actual_delivery).getTime();
                return (act - est) / (1000 * 3600); // hours
            });

        if (delays.length === 0) return { predicted_delay_hours: 0, confidence: 0.6, risk: 'low' };

        // Exponential weighted moving average (alpha=0.3)
        const alpha = 0.3;
        let ewma = delays[0];
        for (let i = 1; i < delays.length; i++) {
            ewma = alpha * delays[i] + (1 - alpha) * ewma;
        }

        const variance = delays.reduce((s, d) => s + Math.pow(d - ewma, 2), 0) / delays.length;
        const stdDev = Math.sqrt(variance);
        const confidence = Math.max(0.3, Math.min(0.95, 1 - (stdDev / (Math.abs(ewma) + 1)) * 0.5));

        return {
            predicted_delay_hours: Math.round(ewma * 10) / 10,
            std_deviation: Math.round(stdDev * 10) / 10,
            confidence: Math.round(confidence * 100) / 100,
            risk: ewma > 24 ? 'high' : ewma > 8 ? 'medium' : 'low',
            samples: delays.length
        };
    }

    /**
     * Inventory forecast using exponential smoothing
     * Predicts stock levels for next N periods
     */
    forecastInventory(history, periodsAhead = 7) {
        if (!history || history.length < 3) {
            return { forecast: [], trend: 'stable', confidence: 0.4, alert: null };
        }

        const values = history.map(h => h.quantity);
        const alpha = 0.4;
        const beta = 0.1;

        // Double exponential smoothing (Holt's method)
        let level = values[0];
        let trend = values.length > 1 ? values[1] - values[0] : 0;

        for (let i = 1; i < values.length; i++) {
            const newLevel = alpha * values[i] + (1 - alpha) * (level + trend);
            trend = beta * (newLevel - level) + (1 - beta) * trend;
            level = newLevel;
        }

        const forecast = [];
        for (let i = 1; i <= periodsAhead; i++) {
            forecast.push({
                period: i,
                predicted: Math.max(0, Math.round(level + trend * i)),
                lower: Math.max(0, Math.round(level + trend * i - values.length * 0.1 * i)),
                upper: Math.round(level + trend * i + values.length * 0.1 * i)
            });
        }

        const lastPrediction = forecast[forecast.length - 1].predicted;
        const trendDirection = trend > 1 ? 'increasing' : trend < -1 ? 'decreasing' : 'stable';

        let alert = null;
        if (lastPrediction < 10) alert = { type: 'understock', message: 'Stock predicted to reach critical low', severity: 'high' };
        else if (lastPrediction > 900) alert = { type: 'overstock', message: 'Stock predicted to exceed capacity', severity: 'medium' };

        return {
            forecast,
            trend: trendDirection,
            confidence: Math.min(0.9, 0.5 + values.length * 0.05),
            current_level: values[values.length - 1],
            alert
        };
    }

    /**
     * Detect supply chain bottlenecks
     * Analyzes throughput at each node
     */
    detectBottlenecks(events, partners) {
        const nodeStats = {};

        (events || []).forEach(e => {
            const key = e.partner_id || e.location || 'unknown';
            if (!nodeStats[key]) {
                nodeStats[key] = { node: key, events: 0, types: {}, avg_dwell: 0, delays: [] };
            }
            nodeStats[key].events++;
            nodeStats[key].types[e.event_type] = (nodeStats[key].types[e.event_type] || 0) + 1;
        });

        // Calculate throughput scores
        const nodes = Object.values(nodeStats);
        if (nodes.length === 0) return { bottlenecks: [], health: 'healthy' };

        const avgEvents = nodes.reduce((s, n) => s + n.events, 0) / nodes.length;

        const bottlenecks = nodes
            .map(n => {
                const ratio = n.events / avgEvents;
                const partnerInfo = (partners || []).find(p => p.id === n.node);
                return {
                    node_id: n.node,
                    node_name: partnerInfo ? partnerInfo.name : n.node,
                    throughput: n.events,
                    throughput_ratio: Math.round(ratio * 100) / 100,
                    event_breakdown: n.types,
                    is_bottleneck: ratio < 0.5 || ratio > 2.0,
                    severity: ratio < 0.3 ? 'critical' : ratio < 0.5 ? 'high' : ratio > 2.0 ? 'warning' : 'normal'
                };
            })
            .sort((a, b) => a.throughput_ratio - b.throughput_ratio);

        const bottleneckCount = bottlenecks.filter(b => b.is_bottleneck).length;

        return {
            bottlenecks,
            health: bottleneckCount === 0 ? 'healthy' : bottleneckCount <= 2 ? 'warning' : 'critical',
            total_nodes: nodes.length,
            bottleneck_count: bottleneckCount
        };
    }

    /**
     * Route optimization — weighted shortest path
     * Considers distance, delay history, and partner risk
     */
    optimizeRoute(graph, fromId, toId) {
        if (!graph || graph.length === 0) return { path: [], cost: 0, optimized: false };

        // Build adjacency list
        const adj = {};
        graph.forEach(edge => {
            if (!adj[edge.from_node_id]) adj[edge.from_node_id] = [];
            adj[edge.from_node_id].push({
                to: edge.to_node_id,
                weight: edge.weight * (1 + edge.risk_score),
                risk: edge.risk_score
            });
        });

        // Dijkstra's algorithm
        const dist = {};
        const prev = {};
        const visited = new Set();
        const queue = [];

        Object.keys(adj).forEach(n => { dist[n] = Infinity; });
        graph.forEach(e => { if (!(e.to_node_id in dist)) dist[e.to_node_id] = Infinity; });
        dist[fromId] = 0;
        queue.push({ node: fromId, cost: 0 });

        while (queue.length > 0) {
            queue.sort((a, b) => a.cost - b.cost);
            const { node } = queue.shift();
            if (visited.has(node)) continue;
            visited.add(node);

            (adj[node] || []).forEach(edge => {
                const newCost = dist[node] + edge.weight;
                if (newCost < dist[edge.to]) {
                    dist[edge.to] = newCost;
                    prev[edge.to] = node;
                    queue.push({ node: edge.to, cost: newCost });
                }
            });
        }

        // Reconstruct path
        const path = [];
        let current = toId;
        while (current && current !== fromId) {
            path.unshift(current);
            current = prev[current];
        }
        if (current === fromId) path.unshift(fromId);

        return {
            path,
            cost: Math.round((dist[toId] || 0) * 100) / 100,
            optimized: path.length > 0 && dist[toId] !== Infinity,
            hops: path.length - 1
        };
    }

    /**
     * Partner risk scoring — composite multi-factor
     */
    scorePartnerRisk(partner, alerts, shipments, violations) {
        let score = 50; // base score

        // Factor 1: KYC status (+20)
        if (partner.kyc_status === 'verified') score += 20;
        else if (partner.kyc_status === 'pending') score += 5;

        // Factor 2: Fraud alert history (-5 per alert)
        const alertCount = (alerts || []).length;
        score -= Math.min(30, alertCount * 5);

        // Factor 3: Delivery reliability (+15 if good)
        const completed = (shipments || []).filter(s => s.status === 'delivered');
        const onTime = completed.filter(s => {
            if (!s.actual_delivery || !s.estimated_delivery) return true;
            return new Date(s.actual_delivery) <= new Date(s.estimated_delivery);
        });
        const reliability = completed.length > 0 ? onTime.length / completed.length : 0.5;
        score += Math.round(reliability * 15);

        // Factor 4: SLA violations (-10 per violation)
        score -= Math.min(20, (violations || []).length * 10);

        // Factor 5: Tenure bonus
        if (partner.created_at) {
            const months = (Date.now() - new Date(partner.created_at).getTime()) / (1000 * 3600 * 24 * 30);
            score += Math.min(10, Math.floor(months));
        }

        score = Math.max(0, Math.min(100, score));

        return {
            score,
            grade: score >= 80 ? 'A' : score >= 60 ? 'B' : score >= 40 ? 'C' : 'D',
            risk_level: score >= 80 ? 'low' : score >= 60 ? 'medium' : score >= 40 ? 'high' : 'critical',
            factors: {
                kyc: partner.kyc_status,
                alert_count: alertCount,
                delivery_reliability: Math.round(reliability * 100) + '%',
                sla_violations: (violations || []).length,
                overall: score
            }
        };
    }

    /**
     * PageRank-style graph analysis for TrustGraph
     */
    pageRank(nodes, edges, iterations = 20, damping = 0.85) {
        const n = nodes.length;
        if (n === 0) return {};

        const nodeIds = nodes.map(n => n.id);
        let ranks = {};
        nodeIds.forEach(id => { ranks[id] = 1 / n; });

        // Build outgoing edge map
        const outgoing = {};
        const incoming = {};
        nodeIds.forEach(id => { outgoing[id] = []; incoming[id] = []; });

        edges.forEach(e => {
            if (outgoing[e.from_node_id]) outgoing[e.from_node_id].push(e.to_node_id);
            if (incoming[e.to_node_id]) incoming[e.to_node_id].push(e.from_node_id);
        });

        for (let iter = 0; iter < iterations; iter++) {
            const newRanks = {};
            nodeIds.forEach(id => {
                let incomingRank = 0;
                incoming[id].forEach(fromId => {
                    const outDegree = outgoing[fromId].length || 1;
                    incomingRank += ranks[fromId] / outDegree;
                });
                newRanks[id] = (1 - damping) / n + damping * incomingRank;
            });
            ranks = newRanks;
        }

        return ranks;
    }

    /**
     * Detect toxic suppliers using graph centrality + risk data
     */
    detectToxicNodes(nodes, edges, alerts) {
        const ranks = this.pageRank(nodes, edges);

        // Calculate betweenness approximation (degree centrality)
        const inDegree = {};
        const outDegree = {};
        nodes.forEach(n => { inDegree[n.id] = 0; outDegree[n.id] = 0; });
        edges.forEach(e => {
            if (inDegree[e.to_node_id] !== undefined) inDegree[e.to_node_id]++;
            if (outDegree[e.from_node_id] !== undefined) outDegree[e.from_node_id]++;
        });

        // Alert count per node
        const alertMap = {};
        (alerts || []).forEach(a => {
            const key = a.partner_id || a.product_id;
            alertMap[key] = (alertMap[key] || 0) + 1;
        });

        return nodes.map(n => {
            const rank = ranks[n.id] || 0;
            const centrality = (inDegree[n.id] + outDegree[n.id]) / Math.max(1, nodes.length - 1);
            const alertCount = alertMap[n.id] || 0;

            // Toxicity = high centrality + high alerts + low trust
            const toxicity = (centrality * 0.3 + (alertCount / 10) * 0.4 + (1 - (n.trust_score || 50) / 100) * 0.3);

            return {
                id: n.id,
                name: n.name,
                type: n.type,
                pagerank: Math.round(rank * 10000) / 10000,
                centrality: Math.round(centrality * 100) / 100,
                in_degree: inDegree[n.id],
                out_degree: outDegree[n.id],
                alert_count: alertCount,
                trust_score: n.trust_score || 50,
                toxicity_score: Math.round(toxicity * 100) / 100,
                is_toxic: toxicity > 0.5,
                risk_level: toxicity > 0.7 ? 'critical' : toxicity > 0.5 ? 'high' : toxicity > 0.3 ? 'medium' : 'low'
            };
        }).sort((a, b) => b.toxicity_score - a.toxicity_score);
    }
}

module.exports = new ScmAI();

File: services/ai-analytics/engines/scm_ai.py
"""
Supply Chain AI Engine
Delay prediction (EWMA), inventory forecast (Holt method), bottleneck detection,
Dijkstra route optimization, partner risk scoring, PageRank, toxic node detection.
Ported from server/engines/scm-ai.js
"""

from __future__ import annotations

import math
import heapq
from collections import defaultdict, Counter
from datetime import datetime, timezone
from typing import Any


def predict_delay(shipments: list[dict]) -> dict:
    """Predict delivery delay using exponential weighted moving average."""
    if not shipments or len(shipments) < 2:
        return {"predicted_delay_hours": 0, "confidence": 0.5, "risk": "low"}

    delays = []
    for s in shipments:
        if s.get("actual_delivery") and s.get("estimated_delivery"):
            try:
                est = datetime.fromisoformat(str(s["estimated_delivery"]).replace("Z", "+00:00")).timestamp()
                act = datetime.fromisoformat(str(s["actual_delivery"]).replace("Z", "+00:00")).timestamp()
                delays.append((act - est) / 3600)
            except Exception:
                pass

    if not delays:
        return {"predicted_delay_hours": 0, "confidence": 0.6, "risk": "low"}

    alpha = 0.3
    ewma = delays[0]
    for d in delays[1:]:
        ewma = alpha * d + (1 - alpha) * ewma

    var = sum((d - ewma) ** 2 for d in delays) / len(delays)
    std = math.sqrt(var)
    conf = max(0.3, min(0.95, 1 - (std / (abs(ewma) + 1)) * 0.5))

    return {
        "predicted_delay_hours": round(ewma, 1),
        "std_deviation": round(std, 1),
        "confidence": round(conf, 2),
        "risk": "high" if ewma > 24 else ("medium" if ewma > 8 else "low"),
        "samples": len(delays),
    }


def forecast_inventory(history: list[dict], periods_ahead: int = 7) -> dict:
    """Inventory forecast using double exponential smoothing (Holt)."""
    if not history or len(history) < 3:
        return {"forecast": [], "trend": "stable", "confidence": 0.4, "alert": None}

    values = [h.get("quantity", 0) for h in history]
    alpha, beta = 0.4, 0.1
    level = values[0]
    trend = values[1] - values[0] if len(values) > 1 else 0

    for v in values[1:]:
        new_level = alpha * v + (1 - alpha) * (level + trend)
        trend = beta * (new_level - level) + (1 - beta) * trend
        level = new_level

    fc = []
    for i in range(1, periods_ahead + 1):
        pred = max(0, round(level + trend * i))
        fc.append({
            "period": i, "predicted": pred,
            "lower": max(0, round(level + trend * i - len(values) * 0.1 * i)),
            "upper": round(level + trend * i + len(values) * 0.1 * i),
        })

    last = fc[-1]["predicted"]
    trend_dir = "increasing" if trend > 1 else ("decreasing" if trend < -1 else "stable")
    alert = None
    if last < 10:
        alert = {"type": "understock", "message": "Stock predicted to reach critical low", "severity": "high"}
    elif last > 900:
        alert = {"type": "overstock", "message": "Stock predicted to exceed capacity", "severity": "medium"}

    return {
        "forecast": fc, "trend": trend_dir,
        "confidence": min(0.9, 0.5 + len(values) * 0.05),
        "current_level": values[-1], "alert": alert,
    }


def detect_bottlenecks(events: list[dict], partners: list[dict]) -> dict:
    """Detect supply chain bottlenecks via throughput analysis."""
    stats: dict[str, dict] = {}
    for e in events or []:
        key = e.get("partner_id") or e.get("location") or "unknown"
        if key not in stats:
            stats[key] = {"node": key, "events": 0, "types": defaultdict(int)}
        stats[key]["events"] += 1
        stats[key]["types"][e.get("event_type", "other")] += 1

    nodes = list(stats.values())
    if not nodes:
        return {"bottlenecks": [], "health": "healthy"}

    avg = sum(n["events"] for n in nodes) / len(nodes)
    partner_map = {p.get("id"): p.get("name") for p in (partners or [])}

    bottlenecks = []
    for n in nodes:
        ratio = n["events"] / avg if avg else 1
        bottlenecks.append({
            "node_id": n["node"], "node_name": partner_map.get(n["node"], n["node"]),
            "throughput": n["events"], "throughput_ratio": round(ratio, 2),
            "event_breakdown": dict(n["types"]),
            "is_bottleneck": ratio < 0.5 or ratio > 2.0,
            "severity": "critical" if ratio < 0.3 else ("high" if ratio < 0.5 else ("warning" if ratio > 2.0 else "normal")),
        })
    bottlenecks.sort(key=lambda b: b["throughput_ratio"])
    bc = sum(1 for b in bottlenecks if b["is_bottleneck"])
    return {"bottlenecks": bottlenecks, "health": "healthy" if bc == 0 else ("warning" if bc <= 2 else "critical"), "total_nodes": len(nodes), "bottleneck_count": bc}


def optimize_route(graph: list[dict], from_id: str, to_id: str) -> dict:
    """Dijkstra shortest path with risk-weighted edges (heapq optimized)."""
    if not graph:
        return {"path": [], "cost": 0, "optimized": False}

    adj: dict[str, list[tuple[str, float]]] = defaultdict(list)
    all_nodes: set[str] = set()
    for e in graph:
        f, t = e.get("from_node_id", ""), e.get("to_node_id", "")
        all_nodes.update((f, t))
        w = e.get("weight", 1) * (1 + e.get("risk_score", 0))
        adj[f].append((t, w))

    dist = {n: float("inf") for n in all_nodes}
    prev: dict[str, str | None] = {}
    dist[from_id] = 0
    visited: set[str] = set()
    # heapq priority queue: (cost, node)
    heap = [(0.0, from_id)]

    while heap:
        cost, node = heapq.heappop(heap)
        if node in visited:
            continue
        if node == to_id:
            break  # Early exit when target reached
        visited.add(node)
        for neighbor, weight in adj[node]:
            nc = cost + weight
            if nc < dist.get(neighbor, float("inf")):
                dist[neighbor] = nc
                prev[neighbor] = node
                heapq.heappush(heap, (nc, neighbor))

    # Path reconstruction: append + reverse (O(n) vs insert(0) O(n²))
    path = []
    cur: str | None = to_id
    while cur and cur != from_id:
        path.append(cur)
        cur = prev.get(cur)
    if cur == from_id:
        path.append(from_id)
    path.reverse()

    return {
        "path": path,
        "cost": round(dist.get(to_id, 0), 2),
        "optimized": len(path) > 0 and dist.get(to_id, float("inf")) != float("inf"),
        "hops": len(path) - 1,
    }


def score_partner_risk(partner: dict, alerts: list[dict], shipments: list[dict], violations: list[dict]) -> dict:
    """Composite multi-factor partner risk score."""
    score = 50
    kyc = partner.get("kyc_status")
    if kyc == "verified":
        score += 20
    elif kyc == "pending":
        score += 5
    ac = len(alerts or [])
    score -= min(30, ac * 5)

    completed = [s for s in (shipments or []) if s.get("status") == "delivered"]
    on_time = [s for s in completed if not s.get("actual_delivery") or not s.get("estimated_delivery") or s["actual_delivery"] <= s["estimated_delivery"]]
    rel = len(on_time) / len(completed) if completed else 0.5
    score += round(rel * 15)
    score -= min(20, len(violations or []) * 10)

    if partner.get("created_at"):
        try:
            created = datetime.fromisoformat(str(partner["created_at"]).replace("Z", "+00:00"))
            months = (datetime.now(timezone.utc) - created).days / 30
            score += min(10, int(months))
        except Exception:
            pass

    score = max(0, min(100, score))
    return {
        "score": score,
        "grade": "A" if score >= 80 else ("B" if score >= 60 else ("C" if score >= 40 else "D")),
        "risk_level": "low" if score >= 80 else ("medium" if score >= 60 else ("high" if score >= 40 else "critical")),
        "factors": {"kyc": kyc, "alert_count": ac, "delivery_reliability": f"{round(rel * 100)}%", "sla_violations": len(violations or []), "overall": score},
    }


def page_rank(nodes: list[dict], edges: list[dict], iterations: int = 20, damping: float = 0.85) -> dict[str, float]:
    """PageRank-style graph analysis (optimized: swap refs instead of new dict per iter)."""
    n = len(nodes)
    if n == 0:
        return {}
    ids = [nd.get("id") for nd in nodes]
    ranks = {nid: 1 / n for nid in ids}
    outgoing: dict[str, list[str]] = {nid: [] for nid in ids}
    incoming: dict[str, list[str]] = {nid: [] for nid in ids}
    # Pre-compute outgoing counts for denominator
    out_count: dict[str, int] = {nid: 0 for nid in ids}

    for e in edges:
        f, t = e.get("from_node_id"), e.get("to_node_id")
        if f in outgoing:
            outgoing[f].append(t)
            out_count[f] += 1
        if t in incoming:
            incoming[t].append(f)

    new_ranks = {nid: 0.0 for nid in ids}
    base = (1 - damping) / n

    for _ in range(iterations):
        for nid in ids:
            ir = sum(ranks[fid] / max(out_count[fid], 1) for fid in incoming[nid])
            new_ranks[nid] = base + damping * ir
        ranks, new_ranks = new_ranks, ranks  # Swap refs — zero allocation
    return ranks


def detect_toxic_nodes(nodes: list[dict], edges: list[dict], alerts: list[dict]) -> list[dict]:
    """Detect toxic suppliers using PageRank + centrality + risk."""
    ranks = page_rank(nodes, edges)
    in_d: Counter = Counter()
    out_d: Counter = Counter()
    for e in edges:
        f, t = e.get("from_node_id"), e.get("to_node_id")
        if t:
            in_d[t] += 1
        if f:
            out_d[f] += 1

    alert_map: Counter = Counter()
    for a in (alerts or []):
        k = a.get("partner_id") or a.get("product_id")
        if k:
            alert_map[k] += 1

    n = max(len(nodes) - 1, 1)
    result = []
    for nd in nodes:
        nid = nd.get("id")
        cent = (in_d[nid] + out_d[nid]) / n
        ac = alert_map[nid]
        tox = cent * 0.3 + (ac / 10) * 0.4 + (1 - nd.get("trust_score", 50) / 100) * 0.3
        result.append({
            "id": nid, "name": nd.get("name"), "type": nd.get("type"),
            "pagerank": round(ranks.get(nid, 0), 4), "centrality": round(cent, 2),
            "in_degree": in_d[nid], "out_degree": out_d[nid],
            "alert_count": ac, "trust_score": nd.get("trust_score", 50),
            "toxicity_score": round(tox, 2), "is_toxic": tox > 0.5,
            "risk_level": "critical" if tox > 0.7 else ("high" if tox > 0.5 else ("medium" if tox > 0.3 else "low")),
        })
    result.sort(key=lambda x: x["toxicity_score"], reverse=True)
    return result
